src/Sampler.cpp

```diff
@@ -0,0 +1,232 @@
+
+    // construct white noise right-hand side
+    // (cell_volume is supposed to represent fine-grid W_h)
+    for (int i = 0; i < fine_vector_size_; ++i)
+    {
+        rhs_fine_(i) = scalar_g_ * std::sqrt(cell_volume_) *
+                       normal_distribution_.Sample();
+    }
+}
+
+void PDESampler::NewCoarseSample()
+{
+    current_state_ = COARSE_SAMPLE;
+    MFEM_ASSERT(false, "Not implemented!");
+}
+
+const mfem::Vector& PDESampler::GetFineCoefficient()
+{
+    MFEM_ASSERT(current_state_ == FINE_SAMPLE,
+                "PDESampler object in wrong state (call NewSample() first)!");
+
+    fvupscale_->SolveFine(rhs_fine_, coefficient_fine_);
+    for (int i = 0; i < coefficient_fine_.Size(); ++i)
+    {
+        coefficient_fine_(i) = std::exp(coefficient_fine_(i));
+    }
+    return coefficient_fine_;
+}
+
+/**
+   Implementation notes:
+
+   c_i              : coefficient for coarse basis function, representing ~normal field K
+   (c_i / q_i)      : value of ~normal field K on agg i
+   exp(c_i/q_i)     : value of lognormal field exp(K) on agg i (what this returns)
+   exp(c_i/q_i) q_i : coefficient for coarse basis function, representing lognormal field exp(K) (what ForVisualization variant returns)
+
+   indexing: we consider only the coefficient for the *constant* component i
+*/
+mfem::Vector& PDESampler::GetCoarseCoefficient()
+{
+    MFEM_ASSERT(current_state_ == FINE_SAMPLE ||
+                current_state_ == COARSE_SAMPLE,
+                "PDESampler object in wrong state (call NewSample() first)!");
+
+    if (current_state_ == FINE_SAMPLE)
+        fvupscale_->Restrict(rhs_fine_, rhs_coarse_);
+    mfem::Vector coarse_sol = fvupscale_->GetCoarseVector();
+    fvupscale_->SolveCoarse(rhs_coarse_, coarse_sol);
+    coarse_sol *= -1.0; // ?
+
+    coefficient_coarse_ = 0.0;
+    mfem::Vector coarse_constant_rep = fvupscale_->GetCoarseConstantRep();
+    MFEM_ASSERT(coarse_constant_rep.Size() == coarse_sol.Size(),
+                "PDESampler::GetCoarseCoefficient : Sizes do not match!");
+    int agg_index = 0;
+    for (int i = 0; i < coarse_sol.Size(); ++i)
+    {
+        if (std::fabs(coarse_constant_rep(i)) > 1.e-8)
+        {
+            coefficient_coarse_(agg_index++) =
+                std::exp(coarse_sol(i) / coarse_constant_rep(i));
+        }
+    }
+    MFEM_ASSERT(agg_index == num_coarse_aggs_, "Something wrong in coarse_constant_rep!");
+
+    return coefficient_coarse_;
+}
+
+mfem::Vector& PDESampler::GetCoarseCoefficientForVisualization()
+{
+    MFEM_ASSERT(current_state_ == FINE_SAMPLE ||
+                current_state_ == COARSE_SAMPLE,
+                "PDESampler object in wrong state (call NewSample() first)!");
+
+    if (current_state_ == FINE_SAMPLE)
+        fvupscale_->Restrict(rhs_fine_, rhs_coarse_);
+    coefficient_coarse_.SetSize(rhs_coarse_.Size());
+    fvupscale_->SolveCoarse(rhs_coarse_, coefficient_coarse_);
+    coefficient_coarse_ *= -1.0; // ??
```

Comment:
I think this `*= -1.0;` needs to go into `Upscale::SolveCoarse` instead.  There may not  be anywhere else where it was used.